django/contrib/auth/views.py

```diff
@@ -438,20 +442,35 @@ class PasswordResetConfirmView(PasswordContextMixin, FormView):
     title = _('Enter new password')
     token_generator = default_token_generator
 
+    def _build_redirect_url(self):
+        current_token = self.request.resolver_match.kwargs['token']
+        return self.request.path.replace(current_token, INTERNAL_RESET_URL_TOKEN)
+
     @method_decorator(sensitive_post_parameters())
     @method_decorator(never_cache)
     def dispatch(self, *args, **kwargs):
         assert 'uidb64' in kwargs and 'token' in kwargs
 
         self.validlink = False
-        self.user = self.get_user(kwargs['uidb64'])
 
-        if self.user is not None and self.token_generator.check_token(self.user, kwargs['token']):
+        # Store the token in the session, to not leak it in a referer
+        # and redirect to a token-free URL.
+        # See https://robots.thoughtbot.com/is-your-site-leaking-password-reset-links for details.
+        token = kwargs['token']
+        if token != INTERNAL_RESET_URL_TOKEN:
```

Comment:
It might be smarter to validate the token first and only modify the session + redirect if it's valid. Otherwise it makes it really easy to create a session just by GET'ing a url (possible DoS vector). It also means you can't pass `accounts/password_reset` as the token and take advantage of our `request.path.replace()` code. It probably means validating the token twice, which is slightly slower. Seems fine to me if an invalid token gets leaked.